packages/build/src/utils/blobs.ts

```diff
@@ -84,29 +101,49 @@ export const scanForBlobs = async function (buildDir: string, packagePath?: stri
 const METADATA_PREFIX = '$'
 const METADATA_SUFFIX = '.json'
 
-/** Given output directory, find all file paths to upload excluding metadata files */
-export const getKeysToUpload = async (blobsDir: string): Promise<string[]> => {
+/**
+ * Returns the blobs that should be uploaded for a given directory tree. The
+ * result is an array with the blob key, the path to its data file, and the
+ * path to its metadata file.
+ */
+export const getKeysToUpload = async (blobsDir: string) => {
+  const blobsToUpload: { key: string; contentPath: string; metadataPath: string }[] = []
   const files = await new fdir()
     .withRelativePaths() // we want the relative path from the blobsDir
     .filter((fpath) => !path.basename(fpath).startsWith(METADATA_PREFIX))
     .crawl(blobsDir)
     .withPromise()
 
-  // normalize the path separators to all use the forward slash
-  return files.map((f) => f.split(path.sep).join('/'))
+  files.forEach((filePath) => {
+    const key = filePath.split(path.sep).join('/')
+    const contentPath = path.resolve(blobsDir, filePath)
+    const basename = path.basename(filePath)
+    const metadataPath = path.resolve(
+      blobsDir,
+      path.dirname(filePath),
+      `${METADATA_PREFIX}${basename}${METADATA_SUFFIX}`,
+    )
+
+    blobsToUpload.push({
+      key,
+      contentPath,
+      metadataPath,
+    })
+  })
+
+  return blobsToUpload
 }
 
 /** Read a file and its metadata file from the blobs directory */
 export const getFileWithMetadata = async (
```

Comment:
Because of the changes we made to `getKeysToUpload`, this method is now simpler and receives the path of the content and metadata files, as opposed to having to compute them itself.